javascript-tests pipeline is failing and needs further investigation

Build: https://drone.owncloud.com/owncloud/contacts/1229/4/4

```console
+ cd /var/www/owncloud/server/apps/contacts
+ make npm
npm run build

> contacts@1.5.4 prebuild /var/www/owncloud/server/apps/contacts
> npm install && npm update && node_modules/bower/bin/bower install && node_modules/bower/bin/bower update

npm WARN deprecated codecov@3.8.3: https://about.codecov.io/blog/codecov-uploader-deprecation-plan/
npm WARN deprecated istanbul@0.4.5: This module is no longer maintained, try this instead:
npm WARN deprecated   npm i nyc
npm WARN deprecated Visit https://istanbul.js.org/integrations for other alternatives.
npm WARN deprecated wreck@6.3.0: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated mkdirp@0.5.1: Legacy versions of mkdirp are no longer supported. Please update to mkdirp 1.x. (Note that the API surface has changed to use Promises in 1.x.)
npm WARN deprecated gulp-util@3.0.8: gulp-util is deprecated - replace it, following the guidelines at https://medium.com/gulpjs/gulp-util-ca3b1f9f9ac5
npm WARN deprecated formatio@1.1.1: This package is unmaintained. Use @sinonjs/formatio instead
npm WARN deprecated samsam@1.1.2: This package has been deprecated in favour of @sinonjs/samsam
npm WARN deprecated chokidar@2.1.8: Chokidar 2 does not receive security updates since 2019. Upgrade to chokidar 3 with 15x fewer dependencies
npm WARN deprecated hoek@2.16.3: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated source-map-resolve@0.5.3: See https://github.com/lydell/source-map-resolve#deprecated
npm WARN deprecated boom@2.10.1: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated samsam@1.1.3: This package has been deprecated in favour of @sinonjs/samsam
npm WARN deprecated request@2.88.0: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated fsevents@1.2.13: fsevents 1 will break on node v14+ and could be using insecure binaries. Upgrade to fsevents 2.
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated source-map-url@0.4.1: See https://github.com/lydell/source-map-url#deprecated
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated circular-json@0.3.3: CircularJSON is in maintenance only, flatted is its successor.

> wd@1.14.0 install /var/www/owncloud/server/apps/contacts/node_modules/wd
> node scripts/build-browser-scripts


> es5-ext@0.10.60 postinstall /var/www/owncloud/server/apps/contacts/node_modules/es5-ext
>  node -e "try{require('./_postinstall')}catch(e){}"

npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.2.7 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@~2.3.2 (node_modules/karma/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

added 889 packages from 1634 contributors and audited 891 packages in 27.622s

47 packages are looking for funding
  run `npm fund` for details

found 31 vulnerabilities (2 low, 12 moderate, 11 high, 6 critical)
  run `npm audit fix` to fix them, or `npm audit` for details
bower ESUDO         Please do not run with sudo

Additional error details:
Since bower is a user command, there is no need to execute it with superuser permissions.
If you're having permission errors when using bower without sudo, please spend a few minutes learning more about how your system should work and make any necessary repairs.

http://www.joyent.com/blog/installing-node-and-npm
https://gist.github.com/isaacs/579814

You can however run a command with sudo using "--allow-root" option
bower ui-select#0.14.9      not-cached https://github.com/angular-ui/ui-select.git#0.14.9
bower ui-select#0.14.9         resolve https://github.com/angular-ui/ui-select.git#0.14.9
bower angular-bootstrap#1.1.2       not-cached https://github.com/angular-ui/bootstrap-bower.git#1.1.2
bower angular-bootstrap#1.1.2          resolve https://github.com/angular-ui/bootstrap-bower.git#1.1.2
bower angular-sanitize#1.5.8        not-cached https://github.com/angular/bower-angular-sanitize.git#1.5.8
bower angular-sanitize#1.5.8           resolve https://github.com/angular/bower-angular-sanitize.git#1.5.8
bower angular-cache#4.6.0           not-cached https://github.com/jmdobry/angular-cache.git#4.6.0
bower angular-cache#4.6.0              resolve https://github.com/jmdobry/angular-cache.git#4.6.0
bower angular-route#1.5.8           not-cached https://github.com/angular/bower-angular-route.git#1.5.8
bower angular-route#1.5.8              resolve https://github.com/angular/bower-angular-route.git#1.5.8
bower ngclipboard#^1.1.1            not-cached https://github.com/sachinchoolur/ngclipboard.git#^1.1.1
bower ngclipboard#^1.1.1               resolve https://github.com/sachinchoolur/ngclipboard.git#^1.1.1
bower angular-mocks#1.5.8           not-cached https://github.com/angular/bower-angular-mocks.git#1.5.8
bower angular-mocks#1.5.8              resolve https://github.com/angular/bower-angular-mocks.git#1.5.8
bower angular#1.5.8                 not-cached https://github.com/angular/bower-angular.git#1.5.8
bower angular#1.5.8                    resolve https://github.com/angular/bower-angular.git#1.5.8
bower underscore#^1.8.3             not-cached https://github.com/jashkenas/underscore.git#^1.8.3
bower underscore#^1.8.3                resolve https://github.com/jashkenas/underscore.git#^1.8.3
bower angular-uuid4#0.3.1           not-cached https://github.com/monicao/angular-uuid4.git#0.3.1
bower angular-uuid4#0.3.1              resolve https://github.com/monicao/angular-uuid4.git#0.3.1
bower vcard#0.2.7                   not-cached https://github.com/Heymdall/vcard.git#0.2.7
bower vcard#0.2.7                      resolve https://github.com/Heymdall/vcard.git#0.2.7
bower jquery-timepicker#883bb2cd94  not-cached https://github.com/fgelinas/timepicker.git#883bb2cd94
bower jquery-timepicker#883bb2cd94     resolve https://github.com/fgelinas/timepicker.git#883bb2cd94
bower angular-cache#4.6.0             download https://github.com/jmdobry/angular-cache/archive/4.6.0.tar.gz
bower angular-bootstrap#1.1.2         download https://github.com/angular-ui/bootstrap-bower/archive/1.1.2.tar.gz
bower ui-select#0.14.9                download https://github.com/angular-ui/ui-select/archive/v0.14.9.tar.gz
bower angular-route#1.5.8             download https://github.com/angular/bower-angular-route/archive/v1.5.8.tar.gz
bower angular-sanitize#1.5.8          download https://github.com/angular/bower-angular-sanitize/archive/v1.5.8.tar.gz
bower underscore#^1.8.3               download https://github.com/jashkenas/underscore/archive/1.13.2.tar.gz
bower ngclipboard#^1.1.1              download https://github.com/sachinchoolur/ngclipboard/archive/1.1.3.tar.gz
bower angular#1.5.8                   download https://github.com/angular/bower-angular/archive/v1.5.8.tar.gz
bower angular-mocks#1.5.8             download https://github.com/angular/bower-angular-mocks/archive/v1.5.8.tar.gz
bower angular-uuid4#0.3.1             download https://github.com/monicao/angular-uuid4/archive/v0.3.1.tar.gz
bower jquery-timepicker#883bb2cd94   short-sha Consider using longer commit SHA to avoid conflicts
bower jquery-timepicker#883bb2cd94    download https://github.com/fgelinas/timepicker/archive/883bb2cd94.tar.gz
bower angular-cache#4.6.0              extract archive.tar.gz
bower angular-bootstrap#1.1.2          extract archive.tar.gz
bower angular-bootstrap#1.1.2         resolved https://github.com/angular-ui/bootstrap-bower.git#1.1.2
bower ui-select#0.14.9                 extract archive.tar.gz
bower ngclipboard#^1.1.1               extract archive.tar.gz
bower angular-uuid4#0.3.1              extract archive.tar.gz
bower angular-sanitize#1.5.8           extract archive.tar.gz
bower angular-route#1.5.8              extract archive.tar.gz
bower angular-mocks#1.5.8              extract archive.tar.gz
bower underscore#^1.8.3                extract archive.tar.gz
bower ngclipboard#^1.1.1              resolved https://github.com/sachinchoolur/ngclipboard.git#1.1.3
bower angular-sanitize#1.5.8          resolved https://github.com/angular/bower-angular-sanitize.git#1.5.8
bower angular#1.5.8                    extract archive.tar.gz
bower clipboard#~1.5.5              not-cached https://github.com/zenorocha/clipboard.js.git#~1.5.5
bower clipboard#~1.5.5                 resolve https://github.com/zenorocha/clipboard.js.git#~1.5.5
bower angular-route#1.5.8             resolved https://github.com/angular/bower-angular-route.git#1.5.8
bower jquery-timepicker#883bb2cd94     extract archive.tar.gz
bower angular-mocks#1.5.8             resolved https://github.com/angular/bower-angular-mocks.git#1.5.8
bower angular-uuid4#0.3.1             resolved https://github.com/monicao/angular-uuid4.git#0.3.1
bower angular-cache#4.6.0             resolved https://github.com/jmdobry/angular-cache.git#4.6.0
bower angular#1.5.8                   resolved https://github.com/angular/bower-angular.git#1.5.8
bower ui-select#0.14.9                mismatch Version declared in the json (0.13.2) is different than the resolved one (0.14.9)
bower ui-select#0.14.9                resolved https://github.com/angular-ui/ui-select.git#0.14.9
bower clipboard#~1.5.5                download https://github.com/zenorocha/clipboard.js/archive/v1.5.16.tar.gz
bower jquery-timepicker#883bb2cd94    resolved https://github.com/fgelinas/timepicker.git#883bb2cd94
bower clipboard#~1.5.5                 extract archive.tar.gz
bower clipboard#~1.5.5                resolved https://github.com/zenorocha/clipboard.js.git#1.5.16
bower underscore#^1.8.3               resolved https://github.com/jashkenas/underscore.git#1.13.2
bower vcard#0.2.7                      ECMDERR Failed to execute "git ls-remote --tags --heads https://github.com/Heymdall/vcard.git", exit code of #128 remote: Support for password authentication was removed on August 13, 2021. Please use a personal access token instead. remote: Please see https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations/ for more information. fatal: Authentication failed for 'https://github.com/Heymdall/vcard.git/'

Additional error details:
remote: Support for password authentication was removed on August 13, 2021. Please use a personal access token instead.
remote: Please see https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations/ for more information.
fatal: Authentication failed for 'https://github.com/Heymdall/vcard.git/'
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! contacts@1.5.4 prebuild: `npm install && npm update && node_modules/bower/bin/bower install && node_modules/bower/bin/bower update`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the contacts@1.5.4 prebuild script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2022-04-19T01_19_39_546Z-debug.log
make: *** [Makefile:120: npm] Error 1
```

--- a/.eslintrc.json
+++ b/.eslintrc.json
@@ -25,12 +25,12 @@
 	},
 	"globals": {
 		"angular": false,
-		"vCard": false,
 		"_": false,
 		"OC": false,
 		"t": false,
 		"n": false,
 		"dav": false,
-		"OCA": false
+		"OCA": false,
+		"vCard": false
 	}
 }

--- a/.gitignore
+++ b/.gitignore
@@ -29,3 +29,6 @@ node_modules/
 
 # drone CI is in .drone.star, do not let someone accidentally commit a local .drone.yml
 .drone.yml
+# IDE Caches
+.idea/
+.vscode/

--- a/Makefile
+++ b/Makefile
@@ -170,6 +170,7 @@ appstore: build
 	README.md \
 	js/public \
 	js/dav/dav.js \
+	js/vcard/vcard.js \
 	js/vendor/angular/angular.js \
 	js/vendor/angular-bootstrap/ui-bootstrap.min.js \
 	js/vendor/angular-bootstrap/ui-bootstrap-tpls.min.js \
@@ -181,7 +182,6 @@ appstore: build
 	js/vendor/jquery-timepicker/jquery.ui.timepicker.js \
 	js/vendor/ngclipboard/dist/ngclipboard.min.js \
 	js/vendor/ui-select/dist/select.js \
-	js/vendor/vcard/src/vcard.js \
 	$(appstore_package_name)
 
 ifdef CAN_SIGN

--- a/bower.json
+++ b/bower.json
@@ -26,8 +26,7 @@
     "angular-uuid4": "0.3.1",
     "jquery-timepicker": "883bb2cd94",
     "ui-select": "angular-ui/ui-select#0.14.9",
-    "ngclipboard": "^1.1.1",
-    "vcard": "0.2.7"
+    "ngclipboard": "^1.1.1"
   },
   "devDependencies": {
     "angular-mocks": "1.5.8",

--- a/gulpfile.js
+++ b/gulpfile.js
@@ -20,7 +20,8 @@ gulp.task('eslint', function() {
 		'js/components/**/*.js',
 		'js/models/**/*.js',
 		'js/services/**/*.js',
-		'js/filters/**/*.js'
+		'js/filters/**/*.js',
+		'js/vcard/vcard.js'
 	])
 		.pipe(eslint())
 		.pipe(eslint.format())
@@ -34,7 +35,8 @@ gulp.task('default', gulp.series('eslint', function() {
 		'js/components/**/*.js',
 		'js/models/**/*.js',
 		'js/services/**/*.js',
-		'js/filters/**/*.js'
+		'js/filters/**/*.js',
+		'js/vcard/vcard.js'
 	])
 		// concat (+sourcemaps)
 		.pipe(sourcemaps.init())

--- a/js/vcard/vcard.js
+++ b/js/vcard/vcard.js
@@ -1,0 +1,303 @@
+var vCard = (function () {
+	var PREFIX = 'BEGIN:VCARD',
+		POSTFIX = 'END:VCARD';
+
+	/**
+	 * Return json representation of vCard
+	 * @param {string} string raw vCard
+	 * @returns {*}
+	 */
+	function parse(string) {
+		var result = {},
+			lines = string.split(/\r\n|\r|\n/),
+			count = lines.length,
+			pieces,
+			key,
+			value,
+			meta,
+			namespace;
+
+		for (var i = 0; i < count; i++) {
+			if (lines[i] == '') {
+				continue;
+			}
+			if (lines[i].toUpperCase() == PREFIX || lines[i].toUpperCase() == POSTFIX) {
+				continue;
+			}
+			var data = lines[i];
+
+			/**
+			 * Check that next line continues current
+			 * @param {number} i
+			 * @returns {boolean}
+			 */
+			var isValueContinued = function (i) {
+				return i + 1 < count && (lines[i + 1][0] == ' ' || lines[i + 1][0] == '\t');
+			};
+			// handle multiline properties (i.e. photo).
+			// next line should start with space or tab character
+			if (isValueContinued(i)) {
+				while (isValueContinued(i)) {
+					data += lines[i + 1].trim();
+					i++;
+				}
+			}
+
+			pieces = data.split(':');
+			key = pieces.shift();
+			value = pieces.join(':');
+			namespace = false;
+			meta = {};
+
+			// meta fields in property
+			if (key.match(/;/)) {
+				key = key
+					.replace(/\\;/g, 'ΩΩΩ')
+					.replace(/\\,/, ',');
+				var metaArr = key.split(';').map(function (item) {
+					return item.replace(/ΩΩΩ/g, ';');
+				});
+				key = metaArr.shift();
+				metaArr.forEach(function (item) {
+					var arr = item.split('=');
+					arr[0] = arr[0].toLowerCase();
+					if (arr[0].length === 0) {
+						return;
+					}
+					if (meta[arr[0]]) {
+						meta[arr[0]].push(arr[1]);
+					} else {
+						meta[arr[0]] = [arr[1]];
+					}
+				});
+			}
+
+			// values with \n
+			value = value
+				.replace(/\\n/g, '\n');
+
+			value = tryToSplit(value);
+
+			// Grouped properties
+			if (key.match(/\./)) {
+				var arr = key.split('.');
+				key = arr[1];
+				namespace = arr[0];
+			}
+
+			var newValue = {
+				value: value
+			};
+			if (Object.keys(meta).length) {
+				newValue.meta = meta;
+			}
+			if (namespace) {
+				newValue.namespace = namespace;
+			}
+
+			if (key.indexOf('X-') !== 0) {
+				key = key.toLowerCase();
+			}
+
+			if (typeof result[key] === 'undefined') {
+				result[key] = [newValue];
+			} else {
+				result[key].push(newValue);
+			}
+
+		}
+
+		return result;
+	}
+
+	var HAS_SEMICOLON_SEPARATOR = /[^\\];|^;/,
+		HAS_COMMA_SEPARATOR = /[^\\],|^,/;
+	/**
+	 * Split value by "," or ";" and remove escape sequences for this separators
+	 * @param {string} value
+	 * @returns {string|string[]
+	 */
+	function tryToSplit(value) {
+		if (value.match(HAS_SEMICOLON_SEPARATOR)) {
+			value = value.replace(/\\,/g, ',');
+			return splitValue(value, ';');
+		} else if (value.match(HAS_COMMA_SEPARATOR)) {
+			value = value.replace(/\\;/g, ';');
+			return splitValue(value, ',');
+		} else {
+			return value
+				.replace(/\\,/g, ',')
+				.replace(/\\;/g, ';');
+		}
+	}
+	/**
+	 * Split vcard field value by separator
+	 * @param {string|string[]} value
+	 * @param {string} separator
+	 * @returns {string|string[]}
+	 */
+	function splitValue(value, separator) {
+		var separatorRegexp = new RegExp(separator);
+		var escapedSeparatorRegexp = new RegExp('\\\\' + separator, 'g');
+		// easiest way, replace it with really rare character sequence
+		value = value.replace(escapedSeparatorRegexp, 'ΩΩΩ');
+		if (value.match(separatorRegexp)) {
+			value = value.split(separator);
+
+			value = value.map(function (item) {
+				return item.replace(/ΩΩΩ/g, separator);
+			});
+		} else {
+			value = value.replace(/ΩΩΩ/g, separator);
+		}
+		return value;
+	}
+
+	var guid = (function() {
+		function s4() {
+			return Math.floor((1 + Math.random()) * 0x10000)
+				.toString(16)
+				.substring(1);
+		}
+		return function() {
+			return s4() + s4() + '-' + s4() + '-' + s4() + '-' +
+				s4() + '-' + s4() + s4() + s4();
+		};
+	})();
+
+	var COMMA_SEPARATED_FIELDS = ['nickname', 'related', 'categories', 'pid'];
+
+	var REQUIRED_FIELDS = ['fn'];
+
+	/**
+	 * Generate vCard representation af object
+	 * @param {*} data
+	 * @param {boolean=} addRequired determine if generator should add required properties (version and uid)
+	 * @returns {string}
+	 */
+	function generate(data, addRequired) {
+		var lines = [PREFIX],
+			line = '';
+
+		if (addRequired && !data.version) {
+			data.version = [{value: '3.0'}];
+		}
+		if (addRequired && !data.uid) {
+			data.uid = [{value: guid()}];
+		}
+
+		var escapeCharacters = function (v) {
+			if (typeof v === 'undefined') {
+				return '';
+			}
+			return v
+				.replace(/\n/g, '\\n')
+				.replace(/;/g, '\\;')
+				.replace(/,/g, '\\,');
+		};
+
+		var escapeTypeCharacters = function(v) {
+			if (typeof v === 'undefined') {
+				return '';
+			}
+			return v
+				.replace(/\n/g, '\\n')
+				.replace(/;/g, '\\;');
+		};
+
+		Object.keys(data).forEach(function (key) {
+			if (!data[key] || typeof data[key].forEach !== 'function') {
+				return;
+			}
+			data[key].forEach(function (value) {
+				// ignore undefined values
+				if (typeof value.value === 'undefined') {
+					return;
+				}
+
+				// ignore empty values (unless it's a required field)
+				if (value.value == '' && REQUIRED_FIELDS.indexOf(key) === -1) {
+					return;
+				}
+
+				// ignore empty array values
+				if (value.value instanceof Array) {
+					var empty = true;
+					for (var i = 0; i < value.value.length; i++) {
+						if (typeof value.value[i] !== 'undefined' && value.value[i] != '') {
+							empty = false;
+							break;
+						}
+					}
+					if (empty) {
+						return;
+					}
+				}
+				line = '';
+
+				// add namespace if exists
+				if (value.namespace) {
+					line += value.namespace + '.';
+				}
+				line += key.indexOf('X-') === 0 ? key : key.toUpperCase();
+
+				// add meta properties
+				if (typeof value.meta === 'object') {
+					Object.keys(value.meta).forEach(function (metaKey) {
+						// values of meta tags must be an array
+						if (typeof value.meta[metaKey].forEach !== 'function') {
+							return;
+						}
+						value.meta[metaKey].forEach(function (metaValue) {
+							if (metaKey.length > 0) {
+								if (metaKey.toUpperCase() === 'TYPE') {
+									// Do not escape the comma when it is the type property. This breaks a lot.
+									line += ';' + escapeCharacters(metaKey.toUpperCase()) + '=' + escapeTypeCharacters(metaValue);
+								} else {
+									line += ';' + escapeCharacters(metaKey.toUpperCase()) + '=' + escapeCharacters(metaValue);
+								}
+							}
+						});
+					});
+				}
+
+				line += ':';
+
+
+
+				if (typeof value.value === 'string') {
+					line += escapeCharacters(value.value);
+				} else {
+					// list-values
+					var separator = COMMA_SEPARATED_FIELDS.indexOf(key) !== -1
+						? ','
+						: ';';
+					line += value.value.map(function (item) {
+						return escapeCharacters(item);
+					}).join(separator);
+				}
+
+				// line-length limit. Content lines
+				// SHOULD be folded to a maximum width of 75 octets, excluding the line break.
+				if (line.length > 75) {
+					var firstChunk = line.substr(0, 75),
+						least = line.substr(75);
+					var splitted = least.match(/.{1,74}/g);
+					lines.push(firstChunk);
+					splitted.forEach(function (chunk) {
+						lines.push(' ' + chunk);
+					});
+				} else {
+					lines.push(line);
+				}
+			});
+		});
+
+		lines.push(POSTFIX);
+		return lines.join('\r\n');
+	}
+	return {
+		parse: parse,
+		generate: generate
+	};
+})();

--- a/karma.conf.js
+++ b/karma.conf.js
@@ -20,7 +20,7 @@ module.exports = function(config) {
 			'js/vendor/ui-select/dist/select.js',
 
 			'js/dav/dav.js',
-			'js/vendor/vcard/src/vcard.js',
+			'js/vcard/vcard.js',
 
 			'js/vendor/underscore/underscore-umd.js',
 			'js/vendor/angular-mocks/angular-mocks.js',

